File: copasi/bindings/R/examples/process_callback.R
#
# This example registers a callback handler, 
# so that when an R interrupt is registered
# the running COPASI task will react immediately 
# to interrupts.
#
# Usage: R --slave -f process_callback.R --args <input-filename>
#


# First we load the COPASI package
dyn.load(paste("COPASI", .Platform$dynlib.ext, sep=""))
source("COPASI.R")
cacheMetaData(1)

dataModel <- CRootContainer_addDatamodel()
args <- commandArgs(trailingOnly = TRUE)
# stop if we don't have any arguments
if (length(args) != 1) {
  stop("Need one argument, the COPASI filename to run.")
}

# load teh model
filename <- args[1]
tryCatch(invisible(dataModel$loadModel(filename)), error = function(e) {
  write(paste("Error while loading the model from file named \"" , filename , "\"."), stderr())
  quit(save = "default", status = 1, runLast = TRUE)
} )

# setup callback with a timeout of 1000
# change this value as needed to setup the maximum
# time allotted before the operation will be stopped (0 to disable)
cb <- RProcessCallback(1000)
    
# get tasks 
taskList <- (dataModel$getTaskList())
i <- 0
# iterate through tasks
while ( i < taskList$size())
{
  task <- dataModel$getTask(i)

  if (is.null(task))
  {
    # skip invalid tasks
    i <- i + 1
    next
  }

  if (!task$isScheduled())
  {
    # skip tasks that are not scheduled
    i <- i + 1
    next
  }

  print(paste("Running Scheduled Task: ",task$getObjectName()," stop anytime using CTRL+C | R interrupt"))
  # set callback 
  task$setCallBack(cb)
  # run task
  task$process(TRUE)
  # reset callback 
  task$clearCallBack()
  i <- i + 1
  
}

print("done...")

File: copasi/bindings/python/examples/process_callback.py
#!/bin/env python

#
# This example registers a callback handler, so that when CTRL+C is pressed
# the running COPASI task is asked to be stopped
#

import COPASI
import sys
import signal


try:
    dataModel = COPASI.CCopasiRootContainer.addDatamodel()
except:
    dataModel = COPASI.CRootContainer.addDatamodel()

class PythonProgress(COPASI.CProcessReport):
    def __init__(self, maxTime=0):
        super(PythonProgress, self).__init__(maxTime)
        self.shouldProceed = True

    def progressItem(self, handle):
        print('progress on: {0} shouldProceed={1}'.format(handle, self.shouldProceed))
        return self.shouldProceed

    def proceed(self):
        return self.shouldProceed

    def askToStop(self):
        print('task should stop soon ...')
        self.shouldProceed = False

# instantiate progress report
progress = PythonProgress().__disown__()

# now setup a handler to ask to stop the processing
def sigint_handler(signum, frame):
    global progress
    progress.askToStop()

signal.signal(signal.SIGINT, sigint_handler)


def simulate_file_with_progress(file_name):

    if not dataModel.loadModel(file_name):
        print("could'nt load COPASI file due to: ")
        print(COPASI.CCopasiMessage.getAllMessageText())
        return

    # go through the processes and run the scheduled one
    for i in range(dataModel.getNumTasks()):
        task = dataModel.getTaskList().get(i)
        if task.isScheduled():
            print('Running Scheduled Task: {0}, stop anytime using CTRL+C'.format( task))
            # set progress support
            task.setCallBack(progress)
            # execute the task
            task.process(True)
            # unset
            task.clearCallBack()


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print('usage: progress_report <cps file>')
        sys.exit(1)
    simulate_file_with_progress(sys.argv[1])